src/padding.jl

```diff
@@ -0,0 +1,205 @@
+"""
+    pad_zeros(x, pad::Tuple; [dims])
+    pad_zeros(x, pad::Int; [dims])
+
+Pad the array `x` with zeros.
+Equivalent to [`pad_constant`](@ref) with the constant equal to 0. 
+"""
+pad_zeros(x::AbstractArray, pad::NTuple{M,Int}; dims=1:M÷2) where M =
+  pad_constant(x, pad, 0; dims=dims)
+
+"""
+    pad_constant(x, pad::Tuple, val=0; [dims])
+    pad_constant(x, pad::Int, val=0; [dims])
+
+Pad the array `x` with the constant value `val`.
+
+`pad` can a tuple of integers `(l1, r1, ..., ln, rn)`
+of some length `2n` that specifies the left and right padding size
+for each of the dimensions in `dims`. If `dims` is not given, 
+it defaults to the first `n` dimensions.
+
+For integer `pad` input instead, it is applied on both sides
+on every dimension in `dims`. In this case, `dims` 
+defaults to the first `ndims(x)-2` dimension 
+(i.e. excludes the channel and batch dimension).
+
+See also [`pad_reflect`](@ref) and [`pad_repeat`](@ref).
+
+```jldoctest
+julia> pad_constant(reshape(1:4, 2, 2), (1, 2, 3, 4), 8)
+5×9 Matrix{Int64}:
+ 8  8  8  8  8  8  8  8  8
+ 8  8  8  1  3  8  8  8  8
+ 8  8  8  2  4  8  8  8  8
+ 8  8  8  8  8  8  8  8  8
+ 8  8  8  8  8  8  8  8  8
+````
+"""
+function pad_constant(x::AbstractArray, pad::NTuple{M,Int}, val=0; 
+                    dims=1:M÷2) where M
+  length(dims) == M÷2 ||
+    throw(ArgumentError("The number of dims should be equal to the number of padding dimensions"))
+  outsize, center = compute_pad_outsize_and_center(size(x), pad, dims)
+  y = fill!(similar(x, eltype(x), outsize), val)
+  y[center...] .= x
+  return y
+end
+
+function compute_pad_outsize_and_center(sz::NTuple{N,Int}, pad, dims) where {N}
+  leftpad, rightpad = pad[1:2:end], pad[2:2:end]
+  
+  outsize = ntuple(N) do i
+              k = findfirst(==(i), dims)
+              if k === nothing
+                return sz[i]
+              else
+                return sz[i] + leftpad[k] + rightpad[k]
+              end
+            end::NTuple{N,Int}
+  
+  leftcorner = ones(Int, N)
+  rightcorner = collect(outsize)
+  for (i, d) in enumerate(dims)
+    leftcorner[d] += leftpad[i]
+    rightcorner[d] -= rightpad[i]
+  end
+  center = ntuple(N) do i
+            leftcorner[i]:rightcorner[i]
+          end 
```

Comment:
We could guarantee this function sees `length(pad) == 2 * N` and then do
```julia
center = broadcast((x,y) -> x .+ y, axes(x), pad[2:2:end])
```
or similar which would be cleaner and nice to guarantee we have padding information for all the dims in one place.

Reply:
I get the gist, but I don't think that code is going to work (besides the fact that there is no variable named `x`)
and I can't think of a way to make it right. Also, moving away from ntuple could cause type instability, I don't advise going in that direction

Reply:
`x` is the array as is common practice in so much code. We can still make use of the `NTuple`, in fact, `pad` would be one.

Reply:
```Julia
pad_constant(x::AbstractArray{T,N}, pad::Int, val = 0; kwargs...) where {T,N} =
  pad_constant(x, ntuple(_ -> pad, 2N), val)
pad_constant(x::AbstractArray{T,N}, pad::Tuple, val = 0; dims = :) where {T,N} =
  pad_constant(x, gen_pad(pad, dims, N), val)

function pad_idx(pad, dims, N)
  is = zip( (2 .* dims) .- 1, (2 .* dims))
end

function gen_pad(pad::NTuple{P,Int}, dims::NTuple{D,Int}, N) where {D,P}
  if P == 2N
    return pad
  elseif P == D
   is = pad_idx(pad, dims, N)
   ps = zeros(Int, 2N)
   for (p,i) in zip(pad, is)
     ps[collect(i)] .= p
   end
   ntuple(i -> ps[i], 2N)
  end
end

# Expects length(pad) == 2M
function pad_constant(x::AbstractArray{T,M}, pad::NTuple{N,Int}, val = 0) where {T,M, N}
  sz, c = size_and_center(x, pad)
  res = fill!(similar(x, sz...), val)
  res[c..., ntuple(_ -> Colon(), M-2)...] = x
  res
end

# pad = (l, r, t, b)
function size_and_center(x, pad::NTuple{N,Int}) where N
  ds = 1:2:N
  dst = ntuple(i -> ds[i], N÷2)
  sz = map(i -> pad[i] + pad[i+1], dst) .+ size(x)
  center = broadcast((x,y) -> x .+ y, axes(x), pad[2:2:end]::NTuple{N÷2,Int})
  sz, center
end
```

Excluding the bit with `gen_pad`, which is not optimised at all (but not too bad either), but other than that, this runs 3-6x faster for small arrays, and discounting allocation times, faster for larger arrays as well (I assume because of inference but unsure).

Reply:
I think we can either replace it now, or I will PR it later

Reply:
can you add it later? because your size_and_center doesn't support dims (e.g. `dims=(1,3)`)  and there are also a few things I don't understand